### PR TITLE
optimized sync

### PR DESCRIPTION
**Pull request by jeroiraz**

### CODE REVIEW

**Review by byo — APPROVED** [review]
Added few comments, no blockers though

File: embedded/store/immustore.go [hunk 1]
```diff
@@ -712,10 +794,57 @@ func (s *ImmuStore) syncBinaryLinking() error {
 }
 
 func (s *ImmuStore) WaitForTx(txID uint64, cancellation <-chan struct{}) error {
-	return s.wHub.WaitFor(txID, cancellation)
+	s.waiteesMutex.Lock()
+
+	if s.waiteesCount == s.maxWaitees {
+		s.waiteesMutex.Unlock()
+		return watchers.ErrMaxWaitessLimitExceeded
```
[the remark this answers is not in the document]
Comment (jeroiraz) [reply embedded/store/immustore.go:801]: we count the number of external waitees, wait for indexing in some operations e.g. ExecAll, wait for transactions for replication...

File: embedded/store/immustore.go [hunk 2]
```diff
@@ -1068,38 +1219,48 @@ func (s *ImmuStore) commit(otx *OngoingTx, expectedHeader *TxHeader, waitForInde
...
 	}
 
-	// will overwrite partially written and uncommitted data
-	committedTxID, committedAlh, committedTxLogSize := s.commitState()
+	s.commitStateRWMutex.Lock()
+	defer s.commitStateRWMutex.Unlock()
+
+	// limit the maximum number of pre-committed transactions
+	if s.synced && s.committedTxID+uint64(s.maxActiveTransactions) <= s.preCommittedTxID {
+		return ErrMaxActiveTransactionsLimitExceeded
```
[the remark this answers is not in the document]
Comment (jeroiraz) [reply embedded/store/immustore.go:1254]: maxActiveTransactions is bounded by the number of ongoingTxs or writers... max commit concurrency determines how many can be concurrently processed (at least the steps which are possible to parallelize)

File: embedded/store/immustore.go [hunk 3]
```diff
@@ -1985,17 +2162,34 @@ func (s *ImmuStore) Sync() error {
...
 	}
 
-	err = s.aht.Sync()
+	err = s.cLog.Flush()
 	if err != nil {
 		return err
 	}
 
-	return s.indexer.Sync()
+	err = s.cLog.Sync()
```
[the remark this answers is not in the document]
Comment (jeroiraz) [reply embedded/store/immustore.go:2181]: yes, it's a good observation. I'll take that into account while working on the appendables in the next PR

File: embedded/store/immustore_test.go [hunk 4]
```diff
@@ -566,18 +569,53 @@ func TestImmudbStoreEdgeCases(t *testing.T) {
...
+
+			err = tx.Set([]byte("key"), nil, []byte("value"))
+			require.NoError(t, err)
+
+			_, err = tx.AsyncCommit()
+			require.ErrorIs(t, err, ErrAlreadyClosed)
+		}()
+
+		// wait for the tx to be waiting for sync to happen
+		time.Sleep(10 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (jeroiraz) [reply embedded/store/immustore_test.go:611]: understand, it won't be due to IO given it's using mocked appendables. It's more about waiting for Commit to be called inside the goroutine

File: pkg/api/schema/schema.proto [hunk 5]
```diff
@@ -549,6 +549,10 @@ message DatabaseNullableSettings {
 	NullableUint32 writeTxHeaderVersion = 20;
 
 	NullableBool autoload = 21;
+
+	NullableUint32 syncFrequency = 22;
```
[the remark this answers is not in the document]
Comment (jeroiraz) [reply pkg/api/schema/schema.proto:553]: defined a message type for Milliseconds

**Pull request merged**